yarn/src/test/scala/org/apache/spark/deploy/yarn/YarnSparkHadoopUtilSuite.scala

```diff
@@ -245,4 +247,55 @@ class YarnSparkHadoopUtilSuite extends SparkFunSuite with Matchers with Logging
       System.clearProperty("SPARK_YARN_MODE")
     }
   }
+
+  test("Obtain tokens For HiveMetastore") {
+    val hadoopConf = new Configuration()
+    hadoopConf.set("hive.metastore.kerberos.principal", "bob")
+    // thrift picks up on port 0 and bails out, without trying to talk to endpoint
+    hadoopConf.set("hive.metastore.uris", "http://localhost:0")
+    val util = new YarnSparkHadoopUtil
+    val e = intercept[InvocationTargetException] {
+      util.obtainTokenForHiveMetastoreInner(hadoopConf, "alice")
+    }
+    assertNestedHiveException(e)
+    // expect exception trapping code to unwind this hive-side exception
+    assertNestedHiveException(intercept[InvocationTargetException] {
+      util.obtainTokenForHiveMetastore(hadoopConf)
+    })
+  }
+
+  def assertNestedHiveException(e: InvocationTargetException): Throwable = {
+    val inner = e.getCause
+    if (inner == null) {
+      fail("No inner cause", e)
+    }
+    if (!inner.isInstanceOf[HiveException]) {
+      fail(s"Not a hive exception", inner)
+    }
+    inner
+  }
+
+  test("handleTokenIntrospectionFailure") {
+    val util = new YarnSparkHadoopUtil
+    // downgraded exceptions
+    util.handleTokenIntrospectionFailure("hive", new ClassNotFoundException("cnfe"))
```

Comment:
I think that because there's really only one exception that's currently interesting, you need more code to implement this "shared policy" approach than just catching the one interesting exception in each call site. It's true that if you need to modify the policy you'd need you'd need to duplicate code (or switch to your current approach), but then do you envision needing to do that? What if the policy for each service needs to be different?

Personally I think that the current approach is a little confusing for someone reading the code (and inconsistent; for example the current code catches `Exception` and then feeds it to a method that matches on `Throwable`), and because the policy is so simple, the sharing argument doesn't justify making the code harder to follow.